railgun/controllers/configuration/secret_utils.go

```diff
@@ -0,0 +1,166 @@
+	if secretNamespace == "" {
+		return ctrl.Result{}, fmt.Errorf("kong can not be configured because the required %s env var is not present", controllers.CtrlNamespaceEnv)
+	}
+
+	// get the configuration secret
+	secret, created, err := getOrCreateConfigSecret(ctx, c, secretNamespace)
+	if err != nil {
+		if errors.IsAlreadyExists(err) {
+			log.Info("kong configuration secret was created elsewhere retrying", "namespace", nsn.Namespace, "ingress", nsn.Name)
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if created {
+		log.Info("kong configuration did not exist, was created successfully", "namespace", nsn.Namespace, "ingress", nsn.Name)
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// before we store configuration data for this Ingress object, ensure that it has our finalizer set
+	if !hasFinalizer(obj, KongIngressFinalizer) {
+		log.Info("finalizer is not set for ingress object, setting it", nsn.Namespace, nsn.Name)
+		finalizers := obj.GetFinalizers()
+		obj.SetFinalizers(append(finalizers, KongIngressFinalizer))
+		if err := c.Update(ctx, obj); err != nil { // TODO: patch here instead of update
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// store the ingress record
+	if err := storeRuntimeObject(ctx, c, secret, obj, nsn); err != nil {
+		if errors.IsConflict(err) {
+			log.Error(err, "object updated while reconcilation was running, retrying", nsn.Namespace, nsn.Name)
+			return ctrl.Result{Requeue: true}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	log.Info("kong secret configuration successfully patched patched", "namespace", nsn.Namespace, "name", controllers.ConfigSecretName)
+	return ctrl.Result{}, nil
+}
+
+// isRuntimeObjectSame indicates whether a runtime.Object you intend to store in the configuration secret is the same as what's already stored.
+// This can be used to decide whether or not an update needs to be performed on the configuration secret.
+func isRuntimeObjectSame(secret *corev1.Secret, obj runtime.Object, nsn types.NamespacedName) (bool, error) {
+	// marshal to YAML to check contents
+	cfg, err := yaml.Marshal(obj)
+	if err != nil {
+		return false, err
+	}
+
+	// check if there's any existing object
+	key := configsecret.KeyFor(obj, nsn)
+	if foundCFG, ok := secret.Data[key]; ok {
+		if bytes.Equal(foundCFG, cfg) {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// storeRuntimeObject stores a runtime.Object in the configuration secret. Callers should re-queue after this completes successfully.
+func storeRuntimeObject(ctx context.Context, c client.Client, secret *corev1.Secret, obj runtime.Object, nsn types.NamespacedName) error {
+	// marshal to YAML for storage
+	cfg, err := yaml.Marshal(obj)
+	if err != nil {
+		return err
+	}
+
+	// patch the secret with the runtime.Object contents
+	key := configsecret.KeyFor(obj, nsn)
+	secret.Data[key] = cfg
+
+	return c.Update(ctx, secret) // TODO: patch here instead of update for perf
```

Comment:
Are you confident that `Update` is concurrency-safe? Not saying it isn't, just double-checking.

Reply:
Yes :+1: